File: internal/protocol/hash.go
package protocol

import (
	"crypto/sha256"
	"encoding/hex"
)

// SHA256Sum 返回 32 字节原始摘要。
func SHA256Sum(b []byte) []byte {
	s := sha256.Sum256(b)
	return s[:]
}

// SHA256Hex 返回 64 字符小写十六进制摘要。
func SHA256Hex(b []byte) string {
	return hex.EncodeToString(SHA256Sum(b))
}

// BlobID 返回内容寻址 id：sha256 十六进制的前 32 个字符（128 bit）。
func BlobID(b []byte) string {
	return SHA256Hex(b)[:32]
}

// IsBlobID 校验 id 是否为合法的 32 字符小写十六进制。
func IsBlobID(id string) bool {
	if len(id) != 32 {
		return false
	}
	for i := 0; i < len(id); i++ {
		c := id[i]
		if (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') {
			continue
		}
		return false
	}
	return true
}

File: packages/protocol-ts/src/hash.ts
import { sha256 } from "@noble/hashes/sha256";
import { bytesToHex, hexToBytes, randomBytes, utf8ToBytes } from "@noble/hashes/utils";

export function utf8(s: string): Uint8Array {
  return utf8ToBytes(s);
}

/** 32 字节原始摘要。 */
export function sha256Sum(data: Uint8Array): Uint8Array {
  return sha256(data);
}

/** 64 字符小写十六进制摘要。 */
export function sha256Hex(data: Uint8Array): string {
  return bytesToHex(sha256(data));
}

/** 内容寻址 id：sha256 十六进制前 32 字符（128 bit）。 */
export function blobId(data: Uint8Array): string {
  return sha256Hex(data).slice(0, 32);
}

export function isBlobId(id: string): boolean {
  return typeof id === "string" && /^[0-9a-f]{32}$/.test(id);
}

export { bytesToHex, hexToBytes, randomBytes };